File: Source/WickedSick.Fayde.Client.NativeEngine/Providers/InheritedDataContextPropertyValueProvider.cs
using System;
using System.Windows.Browser;

namespace WickedSick.Fayde.Client.NativeEngine.Providers
{
    public class InheritedDataContextPropertyValueProvider : PropertyValueProvider, IPropertyChangedListener
    {
        public static DependencyPropertyWrapper DataContextProperty;

        private FrameworkElementNative _Source;

        public InheritedDataContextPropertyValueProvider(DependencyObjectNative @do)
            : base(@do, PropertyPrecedence.InheritedDataContext)
        {
        }

        public override object GetPropertyValue(DependencyPropertyWrapper prop)
        {
            if (_Source == null || prop._ID != DataContextProperty._ID)
                return DependencyObjectNative.UNDEFINED;
            return _Source.GetValue(prop);
        }

        internal void SetDataSource(ScriptObject source)
        {
            var sourceNative = DependencyObjectNative.GetFromScriptObject(source) as FrameworkElementNative;
            if (Nullstone.RefEquals(this._Source, sourceNative))
                return;

            var oldValue = this._Source == null ? DependencyObjectNative.UNDEFINED : this._Source.GetValue(DataContextProperty);
            var newValue = sourceNative == null ? DependencyObjectNative.UNDEFINED : sourceNative.GetValue(DataContextProperty);

            _DetachListener(this._Source);
            this._Source = sourceNative;
            _AttachListener(this._Source);

            if (!Nullstone.Equals(oldValue, newValue))
            {
                var error = HtmlPage.Window.CreateInstance("BError");
                _Object._ProviderValueChanged(_Precedence, DataContextProperty, oldValue, newValue, false, false, false, error);
            }
        }

        private void _AttachListener(FrameworkElementNative sourceNative)
        {
            if (sourceNative == null)
                return;
             sourceNative.SubscribePropertyChanged(DataContextProperty, this);
        }
        private void _DetachListener(FrameworkElementNative sourceNative)
        {
            if (sourceNative == null)
                return;
            sourceNative.UnsubscribePropertyChanged(DataContextProperty, this);
        }

        internal void EmitChanged()
        {
            if (_Source != null)
            {
                var error = HtmlPage.Window.CreateInstance("BError");
                _Object._ProviderValueChanged(_Precedence, DataContextProperty, DependencyObjectNative.UNDEFINED, _Source.GetValue(DataContextProperty), true, false, false, error);
            }
        }
        
        public void PropertyChanged(object sender, PropertyChangedEventArgsNative args)
        {
            var error = HtmlPage.Window.CreateInstance("BError");
            _Object._ProviderValueChanged(_Precedence, args.Property, args.OldValue, args.NewValue, true, false, false, error);
        }
    }
}

File: Source/WickedSick.Fayde.Client.Engine/Javascript/Core/PropertyValueProviders/InheritedDataContextPropertyValueProvider.js
/// <reference path="../../Runtime/Nullstone.js" />
/// <reference path="PropertyValueProvider.js"/>
/// <reference path="Enums.js"/>
/// CODE
/// <reference path="../../Data/PropertyChangedListener.js"/>

(function (Fayde) {
    var _InheritedDataContextPropertyValueProvider = Nullstone.Create("_InheritedDataContextPropertyValueProvider", Fayde._PropertyValueProvider, 1);

    _InheritedDataContextPropertyValueProvider.Instance.Init = function (obj) {
        this.Init$_PropertyValueProvider(obj, _PropertyPrecedence.InheritedDataContext);
        this._Source = null;
    };

    _InheritedDataContextPropertyValueProvider.Instance.GetPropertyValue = function (propd) {
        if (!this._Source || propd._ID !== Fayde.FrameworkElement.DataContextProperty._ID)
            return undefined;
        return this._Source._GetValue(propd);
    };
    _InheritedDataContextPropertyValueProvider.Instance.SetDataSource = function (source) {
        if (Nullstone.RefEquals(this._Source, source))
            return;

        var oldValue = this._Source ? this._Source._GetValue(Fayde.FrameworkElement.DataContextProperty) : undefined;
        var newValue = source ? source._GetValue(Fayde.FrameworkElement.DataContextProperty) : undefined;

        this._DetachListener(this._Source);
        this._Source = source;
        this._AttachListener(this._Source);

        if (!Nullstone.Equals(oldValue, newValue)) {
            var error = new BError();
            this._Object._ProviderValueChanged(this._PropertyPrecedence, Fayde.FrameworkElement.DataContextProperty, oldValue, newValue, false, false, false, error);
        }
    };
    _InheritedDataContextPropertyValueProvider.Instance._AttachListener = function (source) {
        if (source) {
            this._DataContextListener = new Fayde.Data.PropertyChangedListener(source, Fayde.FrameworkElement.DataContextProperty, this, this._SourceDataContextChanged);
            //TODO: Add Handler - Destroyed Event
        }
    };
    _InheritedDataContextPropertyValueProvider.Instance._DetachListener = function (source) {
        if (this._DataContextListener) {
            this._DataContextListener.Detach();
            delete this._DataContextListener;
        }
        if (source) {
            //TODO: Remove Handler - Destroyed Event
        }
    };
    _InheritedDataContextPropertyValueProvider.Instance._SourceDataContextChanged = function (sender, args) {
        var error = new BError();
        this._Object._ProviderValueChanged(this._PropertyPrecedence, args.Property, args.OldValue, args.NewValue, true, false, false, error);
    };
    _InheritedDataContextPropertyValueProvider.Instance.EmitChanged = function () {
        if (this._Source) {
            var error = new BError();
            this._Object._ProviderValueChanged(this._PropertyPrecedence, Fayde.FrameworkElement.DataContextProperty, undefined, this._Source._GetValue(Fayde.FrameworkElement.DataContextProperty), true, false, false, error);
        }
    };

    Fayde._InheritedDataContextPropertyValueProvider = Nullstone.FinishCreate(_InheritedDataContextPropertyValueProvider);
})(Nullstone.Namespace("Fayde"));